File: quicksort/java/src/main/java/QuickSort.java
public class QuickSort {
    public static void sort(int[] input) {
        if(input != null && input.length > 1) {
            quickSort(input, 0, input.length);
        }
    }

    private static void quickSort(int[] input, int begin, int end) {
        if ((end-begin) > 1) {
            int pivotIdx = begin;

            swap(input, pivotIdx, begin);

            pivotIdx = begin;

            for (int i = (begin+1);i<end;i++) {
                if(input[i] < input[begin]) {
                    pivotIdx++;
                    swap(input, pivotIdx, i);
                }
            }

            swap(input, pivotIdx, begin);

            quickSort(input, begin, pivotIdx);
            quickSort(input, pivotIdx+1, end);
        }
    }
    private static void swap(int[] input, int from, int to) {
        if (from!=to) {
            int tmp = input[from];
            input[from] = input[to];
            input[to] = tmp;
        }
    }
}

File: quicksort/python3/quicksort.py
def sort(input):
    if input != None and len(input) > 1:
        _quicksort(input, 0, len(input))

def _quicksort(input, begin, end):
    if (end-begin) > 1:
        pivot = begin
        _swap(input, pivot, begin)
        pivot = begin

        for i in range(begin+1, end):
            if input[i] < input[begin]:
                pivot = pivot+1
                _swap(input, pivot, i)
        
        _swap(input, begin, pivot)

        _quicksort(input, begin, pivot)
        _quicksort(input, pivot+1, end)

def _swap(input, from_idx, to_idx):
    if from_idx != to_idx:
        tmp = input[from_idx]
        input[from_idx] = input[to_idx]
        input[to_idx] = tmp

        